### PR TITLE
ref(hc): Denoralize user email into organization members.

### PR DESCRIPTION
**Pull request by corps**
1.  Make organization_service.add_member idempotent
2.  Remove service usage where using the ORM can be inlined 
3.  Handle edge case on organizationmembermapping service upsert where user has been deleted on the control silo already.
4.  Denormalize user.email into organization member objects to make sorting by it easier and reduce cross silo queries.

### CODE REVIEW

File: src/sentry/api/endpoints/organization_index.py [hunk 1]
```diff
@@ -222,23 +219,24 @@ def post(self, request: Request) -> Response:
                             team=team, organizationmember=om, is_active=True
                         )
 
-                    org_setup_complete.send_robust(
-                        instance=org, user=request.user, sender=self.__class__
-                    )
+                org_setup_complete.send_robust(
```
Comment (corps) [inline src/sentry/api/endpoints/organization_index.py:222]: This logic, which only exists in `getsentry`, works on subscription objects and doesn't belong to an outer transaction anyways, so this is more clarifying.

File: src/sentry/api/endpoints/organization_index.py [hunk 2]
```diff
@@ -222,23 +219,24 @@ def post(self, request: Request) -> Response:
...
-                        "organization.created",
-                        org,
-                        actor_id=request.user.id if request.user.is_authenticated else None,
-                    )
+                analytics.record(
+                    "organization.created",
+                    org,
+                    actor_id=request.user.id if request.user.is_authenticated else None,
+                )
+                om.outbox_for_update().drain_shard(max_updates_to_drain=10)
```
Comment (mikejihbe) [inline src/sentry/api/endpoints/organization_index.py:239]: what is max_updates_to_drain? Seems magic number-y. We have to drain them all, no?

File: src/sentry/api/endpoints/organization_member/requests/invite/index.py [hunk 3]
```diff
@@ -75,16 +74,14 @@ def post(self, request: Request, organization) -> Response:
...
-            organization_id=organization.id,
-            default_org_role=organization.default_role,
-            email=result["email"],
-            role=result["role"],
-            inviter_id=request.user.id,
-            invite_status=InviteStatus.REQUESTED_TO_BE_INVITED.value,
-        )
         with transaction.atomic():
-            om = OrganizationMember.objects.get(id=rpc_org_member.id)
+            om = OrganizationMember.objects.create(
```
Comment (mikejihbe) [inline src/sentry/api/endpoints/organization_member/requests/invite/index.py:78]: maybe i'm missing something, but how does the user accepting the invite get directed to the correct silo without a org member mapping?

File: src/sentry/models/organizationmember.py [hunk 4]
```diff
@@ -223,10 +223,13 @@ class OrganizationMember(Model):
     # Deprecated -- no longer used
     type = BoundedPositiveIntegerField(default=50, blank=True)
 
+    # These attributes are replicated via USER_UPDATE category outboxes for the user object associated with the user_id
+    # when it exists.
     user_is_active = models.BooleanField(
         null=False,
         default=True,
     )
+    user_email = models.CharField(max_length=75, null=True, blank=True)
```
Comment (mikejihbe) [inline src/sentry/models/organizationmember.py:232]: how is this different from the `email` above? Why not just use that?

Comment (mikejihbe) [reply src/sentry/models/organizationmember.py:232]: Leave some commentary here explaining the semantic differences.

**Review by mikejihbe — APPROVED** [review]

File: src/sentry/api/endpoints/organization_index.py (hunk 2 above)
Comment (corps) [reply src/sentry/api/endpoints/organization_index.py]: I do, want to change the way that drainy works.  I content against draining them all, but I also agree this is magic numbery.  I have some pending work to improve draining and simplify it, but I don't want to mix it here.  This is consistent with existing behaviors.

File: tests/sentry/hybrid_cloud/test_organizationmembermapping.py [hunk 5]
```diff
@@ -13,6 +13,22 @@
...
+                organizationmember_id=111111,
+                mapping=RpcOrganizationMemberMappingUpdate(
+                    role=self.organization.default_role,
+                    user_id=10001,
+                    email=None,
+                    inviter_id=None,
+                    invite_status=None,
+                ),
+            )
+            is None
```
Comment (GabeVillalobos) [inline tests/sentry/hybrid_cloud/test_organizationmembermapping.py:29]: This might be silly, but should we add an assertion here that none of the stale user data in the DB has been modified as well?

File: src/sentry/models/organizationmember.py (hunk 4 above)
Comment (corps) [reply src/sentry/models/organizationmember.py:236]: Added comments.

File: src/sentry/migrations/0482_denormalize_user_email.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        migrations.SeparateDatabaseAndState(
+            database_operations=[
+                migrations.RunSQL(
+                    """
+                    ALTER TABLE "sentry_organizationmember" ADD COLUMN "user_email" VARCHAR(75) NULL;
+                    """,
+                    reverse_sql="""
+                    ALTER TABLE "sentry_organizationmember" DROP COLUMN "user_email";
+                    """,
+                    hints={"tables": ["sentry_organizationmember"]},
```
Comment (GabeVillalobos) [inline src/sentry/migrations/0482_denormalize_user_email.py:35]: Should we split this migration into its own PR to ensure that we don't have another potential migration revert scenario?

**Review by GabeVillalobos — APPROVED** [review]
Couple of minor questions but nothing blocking

File: src/sentry/api/endpoints/organization_member/requests/invite/index.py (hunk 3 above)
Comment (corps) [reply src/sentry/api/endpoints/organization_member/requests/invite/index.py:78]: We send out emails (signals attached to org member creation), which contain links that point to AuthLogin (control silo) containing enough information to route (based on OrganizationMapping which will already exist) the RPC calls that complete organization invites.

Don't forget the difference between `OrganizationMapping` (required for RPC routing) and `OrganizationMemberMapping`

Comment (corps) [reply src/sentry/api/endpoints/organization_member/requests/invite/index.py:78]: This endpoint is simply the Region API that creates invites, which triggers emails.

File: tests/sentry/hybrid_cloud/test_organizationmembermapping.py (hunk 5 above)
Comment (corps) [reply tests/sentry/hybrid_cloud/test_organizationmembermapping.py:29]: Sure

Comment (corps) [reply tests/sentry/hybrid_cloud/test_organizationmembermapping.py:29]: Ah, so actually, looking at this again -- the whole reason this test needs to exist *is because* there is a constraint between OrganizationMemberMapping and User.  In fact, adding the test that nothing was insert into the db is making the point; there is nothing wrong with there being an entry in the database *except for the fact that the constraint would normally cause an exception and prevent the upsert from working*.

Technically if we removed the constraint and the test still passed, everything would still work.  Does that make sense?

The important parts of this test are that
1.  Inserting stale users should not throw an exception and
2.  It should not return a mapping object, signaling that the last given inputs are stale.

Whether a row is actually inserted is not necessarily important.

File: src/sentry/migrations/0482_denormalize_user_email.py (hunk 6 above)
Comment (corps) [reply src/sentry/migrations/0482_denormalize_user_email.py:35]: I feel highly confident this change will be safe, and the reversal of this migration will be safe.

Comment (markstory) [reply src/sentry/migrations/0482_denormalize_user_email.py:35]: We never run reverse migrations, nor do we have tooling to do so. Column add operations are generally safe though.

File: src/sentry/models/organizationmember.py (hunk 4 above)
Comment (markstory) [reply src/sentry/models/organizationmember.py:236]: If a user has multiple emails, is this the primary email for the user? Do we have outbox messages for user email change to keep this in sync?

File: src/sentry/services/hybrid_cloud/organizationmember_mapping/impl.py [hunk 7]
```diff
@@ -25,25 +25,34 @@ def upsert_mapping(
...
-            existing.organizationmember_id = organizationmember_id
+                existing.role = mapping.role
+                existing.user_id = mapping.user_id
+                existing.email = mapping.email
+                existing.inviter_id = mapping.inviter_id
+                existing.invite_status = mapping.invite_status
+                existing.organizationmember_id = organizationmember_id
 
-            existing.save()
+                existing.save()
```
Comment (markstory) [inline src/sentry/services/hybrid_cloud/organizationmember_mapping/impl.py:48]: Should we be using `create_or_update` here?
